### PR TITLE
feat bucket replicate: added flag to ignore blocks marked for deletion

### PR DESCRIPTION
**Pull request by FUSAKLA**
Hi, we bumped into some race conditions with `tools bucket replicate`  from bucket with a compactor running, such as replication of a block that just gets deleted by compactor. 
Also, this can save some bandwidth, since it's mostly useless to sync blocks that will be deleted eventually.

Would you be ok accepting this? I'm open to any suggestions., thanks!

## Changes

 - added a new flag `--ignore-marked-for-deletion` to the `thanos bucket replicate` command 

## Verification

 - added tests
 - tested manually

### CODE REVIEW

File: pkg/replicate/scheme.go
Comment (GiedriusS) [inline pkg/replicate/scheme.go:201]: Do we need to have `--ignore-deletion-marks-delay` instead? I wonder if it makes sense :thinking: Mhm, probably not, because we are only copying the blocks here, and they aren't being replaced with one another.

Also, do we need to log `err` if we fail (not due to 404) to check whether the deletion mark exists? Preferably with `debug` log level.

File: pkg/replicate/scheme_test.go
Comment (GiedriusS) [inline pkg/replicate/scheme_test.go:145]: Nit: let's do: `testutil.Equals(t, ...)`.

File: cmd/thanos/tools_bucket.go
Comment (GiedriusS) [inline cmd/thanos/tools_bucket.go:698]: I think this should even be the default TBH, surprised that it is not :smile:

**Review by GiedriusS — COMMENTED** [review]
Please rebase on top of newest `main`, those whitespace changes in docs/... are not needed.

File: cmd/thanos/tools_bucket.go
Comment (FUSAKLA) [reply cmd/thanos/tools_bucket.go:698]: Well I was also a bit surprised  :D 
But didn't want to change the default behavior, but if we agree, I'd be happy to add `--replicated-marked-for-deletion` flag instead. Not sure if anyone could have a use case to replicated blocks that should be deleted.

Another thing is that it does not replicate the deletion marks (see https://github.com/thanos-io/thanos/pull/5118)

File: pkg/replicate/scheme_test.go
Comment (FUSAKLA) [reply pkg/replicate/scheme_test.go]: fixed here https://github.com/thanos-io/thanos/pull/5117/commits/53bab986162fa219d1e0f7761e7b5a8e61545076

File: pkg/replicate/scheme.go
Comment (FUSAKLA) [reply pkg/replicate/scheme.go:201]: Yep, I thought it's enough for replication just the fact that it's marked, didn't think of any use case for the delay here.

Debug log added here https://github.com/thanos-io/thanos/pull/5117/commits/69db471417ea9b6b9d6ef5baf8094da0f3d768cb

**Comment by FUSAKLA** [comment]
Thanks, @GiedriusS for the swift review!

The white space changes were caused by the `make docs`, not sure why. Rolled it back.

**Comment by FUSAKLA** [comment]
Passing through the docs test is a task on it's own I see :smile:

**Review by GiedriusS — APPROVED** [review]
LGTM, let's see for a few days if anyone has any other suggestions

File: cmd/thanos/tools_bucket.go
Comment (GiedriusS) [reply cmd/thanos/tools_bucket.go:698]: Maybe let's keep the old behaviour for now :+1:

File: pkg/replicate/scheme.go
Comment (wiardvanrij) [inline pkg/replicate/scheme.go:203]: Is this an error, error? Or is this more like 'it could be there, which is an acceptable use-case'? - I'm questioning if this then needs to be Debug log level, info or perhaps even being a fatal error then?

**Comment by wiardvanrij** [comment]
Only got one question, other than that, LGTM! :)

File: pkg/replicate/scheme.go
Comment (FUSAKLA) [reply pkg/replicate/scheme.go:203]: Right, the level here is disputable. The semantics is "I was not able to tell if the block is marked as deleted" and the question is what should be the behavior in such case.

This way could happen that we replicate a block marked for deletion even if the flag was enabled.  
I don't have a strong opinion here, maybe it would be safer to bail out completely with an error instead of possibly doing something that configuration tends to forbid.

Comment (FUSAKLA) [reply pkg/replicate/scheme.go:203]: @GiedriusS WDYT?

Comment (GiedriusS) [reply pkg/replicate/scheme.go:203]: Mhm, yeah, maybe let's return an error instead of logging it? I remember there being a discussion about the same exact point but can't find anything now. Something along the lines of "crash early" :thinking: maybe we should add this to our code style document

Comment (FUSAKLA) [reply pkg/replicate/scheme.go]: Ok, agreed changed in https://github.com/thanos-io/thanos/commit/b96326a0c66943a712bbd8189b59dbe8ad264e04 to fail with an error

**Comment by FUSAKLA** [comment]
Not sure why but the PR diff broke GitHub somehow :sweat_smile: 
Getting 500 on the changes page..

**Review by wiardvanrij — APPROVED** [review]
LGTM

**Pull request merged**
